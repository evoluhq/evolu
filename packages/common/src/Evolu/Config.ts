import { ConsoleConfig } from "../Console.js";
import { getOrThrow } from "../Result.js";
import { SimpleName } from "../Type.js";
import type { DbIndexesBuilder } from "./Kysely.js";
import type { AppOwner } from "./Owner.js";

export interface Config extends ConsoleConfig {
  /**
   * The name of the Evolu instance. Evolu is multitenant - it can run multiple
   * instances concurrently. Each instance must have a unique name.
   *
   * The instance name is used as the SQLite database filename for persistent
   * storage, ensuring that database files are separated and invisible to each
   * other.
   *
   * The default value is: `Evolu`.
   *
   * ### Example
   *
   * ```ts
   * // name: getOrThrow(SimpleName.from("MyApp"))
   * ```
   */
  readonly name: SimpleName;

  /**
   * URL for Evolu sync and backup server.
   *
   * The default value is `wss://free.evoluhq.com`.
   */
  readonly syncUrl: string;

  /**
   * URL to reload browser tabs after reset or restore.
   *
   * The default value is `/`.
   */
  readonly reloadUrl: string;

  /**
   * Maximum physical clock drift allowed in ms.
   *
   * The default value is 5 * 60 * 1000 (5 minutes).
   */
  readonly maxDrift: number;

  /**
   * Use the `indexes` option to define SQLite indexes.
   *
   * Table and column names are not typed because Kysely doesn't support it.
   *
   * https://medium.com/@JasonWyatt/squeezing-performance-from-sqlite-indexes-indexes-c4e175f3c346
   *
   * ### Example
   *
   * ```ts
   * const evolu = createEvolu(evoluReactDeps)(Schema, {
   *   indexes: (create) => [
   *     create("todoCreatedAt").on("todo").column("createdAt"),
   *     create("todoCategoryCreatedAt")
   *       .on("todoCategory")
   *       .column("createdAt"),
   *   ],
   * });
   * ```
   */
  readonly indexes?: DbIndexesBuilder;

  /**
   * Initial AppOwner to use when creating Evolu instance. If omitted, a new
   * AppOwner will be generated automatically.
   */
  readonly initialAppOwner?: AppOwner;

  /**
   * Use in-memory SQLite database instead of persistent storage. Useful for
   * testing or temporary data that doesn't need persistence.
   *
   * In-memory databases exist only in RAM and are completely destroyed when the
   * process ends, making them forensically safe for sensitive data.
   *
   * The default value is: `false`.
   */
  readonly inMemory?: boolean;
}

export interface ConfigDep {
  readonly config: Config;
}

export const defaultConfig: Config = {
  name: getOrThrow(SimpleName.fromParent("Evolu")),
  syncUrl: "wss://free.evoluhq.com",
  reloadUrl: "/",
  maxDrift: 5 * 60 * 1000,
  enableLogging: false,
};
