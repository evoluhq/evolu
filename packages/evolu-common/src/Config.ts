import * as Context from "effect/Context";
import * as Layer from "effect/Layer";
import * as LogLevel from "effect/LogLevel";
import * as Logger from "effect/Logger";
import * as ManagedRuntime from "effect/ManagedRuntime";
import * as Match from "effect/Match";

export interface Config {
  /**
   * URL to reload browser tabs after {@link Owner} reset or restore.
   *
   * The default value is `/`.
   */
  reloadUrl: string;

  /**
   * URL for Evolu sync and backup server.
   *
   * The default value is `https://evolu.world`.
   */
  syncUrl: string;

  /**
   * Evolu is multitenant - it can run more instances concurrently. Every Evolu
   * instance has to have its own unique name. Database files are separated and
   * invisible to each other.
   *
   * The default value is: `Evolu`.
   */
  name: string;

  /**
   * Maximum physical clock drift allowed in ms.
   *
   * The default value is 5 * 60 * 1000 (5 minutes).
   */
  maxDrift: number;

  /**
   * Setting the minimum log level. The default value is `none`.
   *
   * For development, use `trace` to log all events and `debug` to log only
   * events with values. For production, use `warning`.
   */
  minimumLogLevel: "none" | "trace" | "debug" | "warning";
}

export const Config = Context.GenericTag<Config>("Config");

export const defaultConfig: Config = {
  reloadUrl: "/",
  syncUrl: "https://evolu.world",
  name: "Evolu",
  maxDrift: 5 * 60 * 1000,
  minimumLogLevel: "none",
};

/** https://effect.website/docs/guides/runtime */
export const createRuntime = (
  partialConfig?: Partial<Config>,
): ManagedRuntime.ManagedRuntime<Config, never> => {
  const config = { ...defaultConfig, ...partialConfig };
  const ConfigLive = Layer.succeed(Config, config);

  const minimumLogLevel = Match.value(config.minimumLogLevel).pipe(
    Match.when("debug", () => LogLevel.Debug),
    Match.when("none", () => LogLevel.None),
    Match.when("trace", () => LogLevel.Trace),
    Match.when("warning", () => LogLevel.Warning),
    Match.exhaustive,
  );

  const evoluLayer =
    config.minimumLogLevel === "none"
      ? ConfigLive
      : Layer.mergeAll(
          ConfigLive,
          Logger.minimumLogLevel(minimumLogLevel),
          Logger.replace(Logger.defaultLogger, makeEvoluLogger(config.name)),
        );

  return ManagedRuntime.make(evoluLayer);
};

// TODO: Spans and Effect native variadic log args.
const makeEvoluLogger = (name: string) =>
  Logger.make(({ logLevel, message, date }) => {
    const fn =
      logLevel._tag === "Warning"
        ? "warn"
        : logLevel._tag === "Error"
          ? "error"
          : "log";

    const hours = date.getHours().toString().padStart(2, "0");
    const minutes = date.getMinutes().toString().padStart(2, "0");
    const seconds = date.getSeconds().toString().padStart(2, "0");
    const milliseconds = date.getMilliseconds().toString().padStart(3, "0");
    const formattedDate = `${hours}:${minutes}:${seconds}:${milliseconds}`;

    const messages: unknown[] = Array.isArray(message) ? message : [message];
    globalThis.console[fn](`${formattedDate} [${name}]`, ...messages);
  });
