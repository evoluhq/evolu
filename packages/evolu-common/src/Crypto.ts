import * as Schema from "@effect/schema/Schema";
import { secretbox } from "@noble/ciphers/salsa";
import { concatBytes } from "@noble/ciphers/utils";
import { hmac } from "@noble/hashes/hmac";
import { sha512 } from "@noble/hashes/sha512";
import { randomBytes } from "@noble/hashes/utils";
import { Brand, Context, Effect, Layer } from "effect";
import { customAlphabet, nanoid } from "nanoid";

export interface Bip39 {
  readonly make: Effect.Effect<never, never, Mnemonic>;

  readonly toSeed: (
    mnemonic: Mnemonic,
  ) => Effect.Effect<never, never, Uint8Array>;

  readonly parse: (
    mnemonic: string,
  ) => Effect.Effect<never, InvalidMnemonicError, Mnemonic>;
}

export const Bip39 = Context.Tag<Bip39>("evolu/Bip39");

export interface InvalidMnemonicError {
  readonly _tag: "InvalidMnemonicError";
}

/**
 * Mnemonic is a password generated by Evolu in BIP39 format.
 *
 * A mnemonic, also known as a "seed phrase," is a set of 12 words in a
 * specific order chosen from a predefined list. The purpose of the BIP39
 * mnemonic is to provide a human-readable way of storing a private key.
 */
export type Mnemonic = string & Brand.Brand<"Mnemonic">;

export interface NanoId {
  readonly nanoid: Effect.Effect<never, never, string>;
  readonly nanoidAsNodeId: Effect.Effect<never, never, NodeId>;
}

export const NanoId = Context.Tag<NanoId>("evolu/NanoId");

export const NodeId: Schema.BrandSchema<
  string,
  string & Brand.Brand<"NodeId">
> = Schema.string.pipe(Schema.pattern(/^[\w-]{16}$/), Schema.brand("NodeId"));
export type NodeId = Schema.Schema.To<typeof NodeId>;

const nanoidForNodeId = customAlphabet("0123456789abcdef", 16);

export const NanoIdLive = Layer.succeed(
  NanoId,
  NanoId.of({
    nanoid: Effect.sync(() => nanoid()),
    nanoidAsNodeId: Effect.sync(() => nanoidForNodeId() as NodeId),
  }),
);

/**
 * SLIP-21 implementation
 * https://github.com/satoshilabs/slips/blob/master/slip-0021.md
 */
export const slip21Derive = (
  seed: Uint8Array,
  path: ReadonlyArray<string>,
): Effect.Effect<never, never, Uint8Array> =>
  Effect.sync(() => {
    let m = hmac(sha512, "Symmetric key seed", seed);
    for (let i = 0; i < path.length; i++) {
      const p = new TextEncoder().encode(path[i]);
      const e = new Uint8Array(p.byteLength + 1);
      e[0] = 0;
      e.set(p, 1);
      m = hmac(sha512, m.slice(0, 32), e);
    }
    return m.slice(32, 64);
  });

/**
 * Alias to xsalsa20poly1305, for compatibility with libsodium / nacl
 */
export interface SecretBox {
  readonly seal: (
    key: Uint8Array,
    plaintext: Uint8Array,
  ) => Effect.Effect<never, never, Uint8Array>;

  readonly open: (
    key: Uint8Array,
    ciphertext: Uint8Array,
  ) => Effect.Effect<never, never, Uint8Array>;
}

export const SecretBox = Context.Tag<SecretBox>("evolu/SecretBox");

export const SecretBoxLive = Layer.succeed(
  SecretBox,
  SecretBox.of({
    seal: (key, plaintext) =>
      Effect.sync(() => {
        const nonce = randomBytes(24);
        const ciphertext = secretbox(key, nonce).seal(plaintext);
        return concatBytes(nonce, ciphertext);
      }),
    open: (key, ciphertext) =>
      Effect.sync(() => {
        const nonce = ciphertext.subarray(0, 24);
        const ciphertextWithoutNonce = ciphertext.subarray(24);
        return secretbox(key, nonce).open(ciphertextWithoutNonce);
      }),
  }),
);
