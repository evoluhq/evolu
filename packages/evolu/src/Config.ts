import { Config } from "./Types.js";

const defaultConfig: Config = {
  syncUrl: "https://evolu.world",
  maxDrift: 5 * 60 * 1000,
  reloadUrl: "/",
};

export const createConfig = (config?: Partial<Config>): Config => ({
  ...defaultConfig,
  ...config,
});
