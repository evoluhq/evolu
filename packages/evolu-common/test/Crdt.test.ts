import * as Schema from "@effect/schema/Schema";
import { Context, Effect, Either, pipe } from "effect";
import fs from "fs";
import { describe, expect, test } from "vitest";
import { Config, ConfigLive } from "../src/Config.js";
import {
  AllowedTimeRange,
  Millis,
  Time,
  Timestamp,
  TimestampCounterOverflowError,
  TimestampDriftError,
  TimestampTimeOutOfRangeError,
  diffMerkleTrees,
  initialMerkleTree,
  initialMillis,
  insertIntoMerkleTree,
  makeInitialTimestamp,
  makeSyncTimestamp,
  millisToMerkleTreePath,
  receiveTimestamp,
  sendTimestamp,
  timestampToHash,
  timestampToString,
  unsafeTimestampFromString,
} from "../src/Crdt.js";
import { NanoId, NodeId } from "../src/Crypto.js";
import { makeNode1Timestamp, makeNode2Timestamp } from "./testUtils.js";

test("InitialTimestampLive", () => {
  const timestamp = makeInitialTimestamp.pipe(
    Effect.provideService(
      NanoId,
      NanoId.of({
        nanoid: Effect.succeed("nanoid"),
        nanoidAsNodeId: Effect.succeed("nanoidAsNodeId" as NodeId),
      }),
    ),
    Effect.runSync,
  );
  expect(timestamp.counter).toBe(0);
  expect(timestamp.millis).toBe(860934420000);
  expect(timestamp.node).toBe("nanoidAsNodeId");
});

test("createSyncTimestamp", () => {
  const ts = makeSyncTimestamp();
  expect(ts.counter).toBe(0);
  expect(ts.millis).toBe(860934420000);
  expect(ts.node).toBe("0000000000000000");
});

test("timestampToString", () => {
  expect(pipe(makeSyncTimestamp(), timestampToString)).toMatchInlineSnapshot(
    '"1997-04-13T12:27:00.000Z-0000-0000000000000000"',
  );
});

test("timestampFromString", () => {
  const t = makeSyncTimestamp();
  expect(t).toEqual(pipe(t, timestampToString, unsafeTimestampFromString));
});

test("timestampToHash", () => {
  expect(timestampToHash(makeSyncTimestamp())).toMatchInlineSnapshot(
    "512511670",
  );
});

const config = Config.pipe(Effect.provide(ConfigLive()), Effect.runSync);

const makeMillis = (millis: number): Millis => Schema.parseSync(Millis)(millis);

const context0 = pipe(
  Context.empty(),
  Context.add(Config, config),
  Context.add(Time, { now: Effect.succeed(initialMillis) }),
);

const context1 = pipe(
  Context.empty(),
  Context.add(Config, config),
  Context.add(Time, { now: Effect.succeed(makeMillis(initialMillis + 1)) }),
);

describe("sendTimestamp", () => {
  test("should send monotonically with a monotonic clock", () => {
    expect(
      pipe(
        makeSyncTimestamp(),
        sendTimestamp,
        Effect.provide(context1),
        Effect.runSync,
      ),
    ).toMatchSnapshot();
  });

  test("should send monotonically with a stuttering clock", () => {
    expect(
      pipe(
        makeSyncTimestamp(),
        sendTimestamp,
        Effect.provide(context0),
        Effect.runSync,
      ),
    ).toMatchSnapshot();
  });

  test("should send monotonically with a regressing clock", () => {
    expect(
      pipe(
        makeSyncTimestamp(makeMillis(initialMillis + 1)),
        sendTimestamp,
        Effect.provide(context0),
        Effect.runSync,
      ),
    ).toMatchSnapshot();
  });

  test("should fail with counter overflow", () => {
    let timestamp: Either.Either<
      | TimestampDriftError
      | TimestampCounterOverflowError
      | TimestampTimeOutOfRangeError,
      Timestamp
    > = Either.right(makeSyncTimestamp());

    for (let i = 0; i < 65536; i++) {
      timestamp = pipe(
        timestamp,
        Effect.flatMap(sendTimestamp),
        Effect.map(Either.right),
        Effect.catchAll((e) => Effect.succeed(Either.left(e))),
        Effect.provide(context0),
        Effect.runSync,
      );
    }

    expect(timestamp).toMatchSnapshot();
  });

  test("should fail with clock drift", () => {
    expect(
      pipe(
        makeSyncTimestamp(makeMillis(initialMillis + config.maxDrift + 1)),
        sendTimestamp,
        Effect.catchAll((e) => Effect.succeed(Either.left(e))),
        Effect.provide(context0),
        Effect.runSync,
      ),
    ).toMatchSnapshot();
  });
});

describe("receiveTimestamp", () => {
  test("wall clock is later than both the local and remote timestamps", () => {
    expect(
      pipe(
        receiveTimestamp({
          local: makeNode1Timestamp(),
          remote: makeNode2Timestamp(0, 0),
        }),
        Effect.provide(context1),
        Effect.runSync,
      ),
    ).toMatchSnapshot();
  });

  describe("wall clock is somehow behind", () => {
    test("for the same timestamps millis, we take the bigger counter", () => {
      expect(
        pipe(
          receiveTimestamp({
            local: makeNode1Timestamp(1, 0),
            remote: makeNode2Timestamp(1, 1),
          }),
          Effect.provide(context1),
          Effect.runSync,
        ),
      ).toMatchSnapshot();

      expect(
        pipe(
          receiveTimestamp({
            local: makeNode1Timestamp(1, 1),
            remote: makeNode2Timestamp(1, 0),
          }),
          Effect.provide(context0),
          Effect.runSync,
        ),
      ).toMatchSnapshot();
    });

    test("local millis is later than remote", () => {
      expect(
        pipe(
          receiveTimestamp({
            local: makeNode1Timestamp(2),
            remote: makeNode2Timestamp(1),
          }),
          Effect.provide(context0),
          Effect.runSync,
        ),
      ).toMatchSnapshot();
    });

    test("remote millis is later than local", () => {
      expect(
        pipe(
          receiveTimestamp({
            local: makeNode1Timestamp(1),
            remote: makeNode2Timestamp(2),
          }),
          Effect.provide(context0),
          Effect.runSync,
        ),
      ).toMatchSnapshot();
    });
  });

  test("TimestampDuplicateNodeError", () => {
    expect(
      pipe(
        receiveTimestamp({
          local: makeNode1Timestamp(),
          remote: makeNode1Timestamp(),
        }),
        Effect.catchAll((e) => Effect.succeed(Either.left(e))),
        Effect.provide(context1),
        Effect.runSync,
      ),
    ).toMatchSnapshot();
  });

  test("should fail with clock drift", () => {
    expect(
      pipe(
        receiveTimestamp({
          local: makeSyncTimestamp(
            makeMillis(initialMillis + config.maxDrift + 1),
          ),
          remote: makeNode2Timestamp(),
        }),
        Effect.catchAll((e) => Effect.succeed(Either.left(e))),
        Effect.provide(context0),
        Effect.runSync,
      ),
    ).toMatchSnapshot();

    expect(
      pipe(
        receiveTimestamp({
          local: makeNode2Timestamp(),
          remote: makeSyncTimestamp(
            makeMillis(initialMillis + config.maxDrift + 1),
          ),
        }),
        Effect.catchAll((e) => Effect.succeed(Either.left(e))),
        Effect.provide(context0),
        Effect.runSync,
      ),
    ).toMatchSnapshot();
  });
});

const node1TimestampStart = makeNode1Timestamp();
const node2Timestamp2022 = makeNode2Timestamp(1656873738591);

const someMerkleTree = pipe(
  initialMerkleTree,
  insertIntoMerkleTree(node2Timestamp2022),
);

test("createInitialMerkleTree", () => {
  expect(initialMerkleTree).toMatchSnapshot();
});

describe("insertIntoMerkleTree", () => {
  test("node1TimestampStart", () => {
    expect(
      insertIntoMerkleTree(node1TimestampStart)(initialMerkleTree),
    ).toMatchSnapshot();
  });

  test("node2Timestamp2022", () => {
    expect(
      insertIntoMerkleTree(node2Timestamp2022)(initialMerkleTree),
    ).toMatchSnapshot();
  });

  test("node1TimestampStart then node2Timestamp2022", () => {
    expect(
      pipe(
        initialMerkleTree,
        insertIntoMerkleTree(node1TimestampStart),
        insertIntoMerkleTree(node2Timestamp2022),
      ),
    ).toMatchSnapshot();
  });

  test("the order does not matter", () => {
    expect(
      pipe(
        initialMerkleTree,
        insertIntoMerkleTree(node1TimestampStart),
        insertIntoMerkleTree(node2Timestamp2022),
      ),
    ).toEqual(
      pipe(
        initialMerkleTree,
        insertIntoMerkleTree(node2Timestamp2022),
        insertIntoMerkleTree(node1TimestampStart),
      ),
    );

    const merkle0 = fs.readFileSync("./test/fixtures/merkle0.json", "utf8");
    const merkle1 = fs.readFileSync("./test/fixtures/merkle1.json", "utf8");
    const merkle2 = fs.readFileSync("./test/fixtures/merkle2.json", "utf8");
    const merkle3 = fs.readFileSync("./test/fixtures/merkle3.json", "utf8");

    expect(merkle0).toEqual(merkle1);
    expect(merkle1).toEqual(merkle2);
    expect(merkle2).toEqual(merkle3);
  });
});

describe("diffMerkleTrees", () => {
  test("diff for two initial Merkle Trees", () => {
    expect(
      diffMerkleTrees(initialMerkleTree, initialMerkleTree),
    ).toMatchSnapshot();
  });

  test("diff for initialMerkleTree and mt1", () => {
    expect(
      diffMerkleTrees(initialMerkleTree, someMerkleTree),
    ).toMatchSnapshot();
  });

  test("diff for mt1 and initialMerkleTree", () => {
    expect(
      diffMerkleTrees(someMerkleTree, initialMerkleTree),
    ).toMatchSnapshot();
  });

  test("minute window", () => {
    expect(
      diffMerkleTrees(
        someMerkleTree,
        pipe(
          initialMerkleTree,
          insertIntoMerkleTree(makeNode2Timestamp(60000 - 1)),
        ),
      ),
    ).toMatchSnapshot();

    expect(
      diffMerkleTrees(
        someMerkleTree,
        pipe(
          initialMerkleTree,
          insertIntoMerkleTree(makeNode2Timestamp(60000)),
        ),
      ),
    ).toMatchSnapshot();
  });

  const twentyYears = 1000 * 60 * 60 * 24 * 365 * 20;

  test("find the most recent time when trees were the same", () => {
    // May 17 2023 12:09:55
    const now = 1684318195723 - initialMillis;

    const t1 = pipe(
      initialMerkleTree,
      insertIntoMerkleTree(makeNode1Timestamp(now)),
      insertIntoMerkleTree(makeNode1Timestamp(now + 10000)),
    );
    const t2 = pipe(
      initialMerkleTree,
      insertIntoMerkleTree(makeNode1Timestamp(now)),
    );
    expect(diffMerkleTrees(t1, t2)).toMatchInlineSnapshot(`
      {
        "_id": "Option",
        "_tag": "Some",
        "value": 1684318140000,
      }
    `);

    const t3 = pipe(
      initialMerkleTree,
      insertIntoMerkleTree(makeNode1Timestamp(now + twentyYears)),
      insertIntoMerkleTree(makeNode1Timestamp(now + twentyYears + 10000)),
    );
    const t4 = pipe(
      initialMerkleTree,
      insertIntoMerkleTree(makeNode1Timestamp(now + twentyYears)),
    );
    expect(diffMerkleTrees(t3, t4)).toMatchInlineSnapshot(`
      {
        "_id": "Option",
        "_tag": "Some",
        "value": 2315038140000,
      }
    `);
  });

  // TODO: Make this test deterministic.
  test("sync", () => {
    const startTime = 0;
    const endTime = twentyYears;

    const getRandomTime = (): number => {
      const randomTime = Math.random() * (endTime - startTime) + startTime;
      return new Date(randomTime).getTime();
    };

    const randomTimes: Array<number> = [];
    for (let i = 0; i < 1000; i++) {
      randomTimes.push(getRandomTime());
    }
    randomTimes.sort((a, b) => a - b);

    const timestamps = randomTimes.map((time) =>
      (Math.random() >= 0.5 ? makeNode1Timestamp : makeNode2Timestamp)(time),
    );

    const db1: Array<Timestamp> = [];
    let t1 = initialMerkleTree;
    const addTo1 = (timestamp: Timestamp): void => {
      if (!db1.some((t) => t.millis === timestamp.millis)) {
        db1.push(timestamp);
        t1 = insertIntoMerkleTree(timestamp)(t1);
      }
    };

    const db2: Array<Timestamp> = [];
    let t2 = initialMerkleTree;
    const addTo2 = (timestamp: Timestamp): void => {
      if (!db2.some((t) => t.millis === timestamp.millis)) {
        db2.push(timestamp);
        t2 = insertIntoMerkleTree(timestamp)(t2);
      }
    };

    const getRandomNumber = (min: number, max: number): number => {
      return Math.random() * (max - min) + min;
    };

    while (timestamps.length > 0) {
      timestamps.splice(0, getRandomNumber(0, 10)).forEach((timestamp) => {
        if (timestamp.node === "0000000000000001") addTo1(timestamp);
        else addTo2(timestamp);
      });

      // eslint-disable-next-line no-constant-condition
      while (1) {
        const diff = diffMerkleTrees(t1, t2);
        if (diff._tag === "None") break;
        [...db1, ...db2]
          .filter((t) => t.millis >= diff.value)
          .forEach((timestamp) => {
            addTo1(timestamp);
            addTo2(timestamp);
          });
      }
    }

    expect(timestamps.length).toBe(0);
  });
});

test("millisToMinutesBase3", () => {
  expect(
    millisToMerkleTreePath(AllowedTimeRange.greaterThan as Millis).length,
  ).toEqual(15);
  expect(
    millisToMerkleTreePath((AllowedTimeRange.greaterThan + 1) as Millis).length,
  ).toEqual(16);
  expect(
    millisToMerkleTreePath((AllowedTimeRange.lessThan - 1) as Millis).length,
  ).toEqual(16);
  expect(
    millisToMerkleTreePath(AllowedTimeRange.lessThan as Millis).length,
  ).toEqual(17);
});
