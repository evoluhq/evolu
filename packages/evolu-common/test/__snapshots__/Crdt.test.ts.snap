// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createInitialMerkleTree 1`] = `{}`;

exports[`diffMerkleTrees > diff for initialMerkleTree and mt1 1`] = `
{
  "_id": "Option",
  "_tag": "Some",
  "value": 860934420000,
}
`;

exports[`diffMerkleTrees > diff for mt1 and initialMerkleTree 1`] = `
{
  "_id": "Option",
  "_tag": "Some",
  "value": 860934420000,
}
`;

exports[`diffMerkleTrees > diff for two initial Merkle Trees 1`] = `
{
  "_id": "Option",
  "_tag": "None",
}
`;

exports[`diffMerkleTrees > minute window 1`] = `
{
  "_id": "Option",
  "_tag": "Some",
  "value": 860934420000,
}
`;

exports[`diffMerkleTrees > minute window 2`] = `
{
  "_id": "Option",
  "_tag": "Some",
  "value": 860934420000,
}
`;

exports[`insertIntoMerkleTree > node1TimestampStart 1`] = `
{
  "1": {
    "0": {
      "0": {
        "0": {
          "0": {
            "0": {
              "0": {
                "0": {
                  "0": {
                    "0": {
                      "0": {
                        "0": {
                          "0": {
                            "0": {
                              "0": {
                                "0": {
                                  "hash": -1854333397,
                                },
                                "hash": -1854333397,
                              },
                              "hash": -1854333397,
                            },
                            "hash": -1854333397,
                          },
                          "hash": -1854333397,
                        },
                        "hash": -1854333397,
                      },
                      "hash": -1854333397,
                    },
                    "hash": -1854333397,
                  },
                  "hash": -1854333397,
                },
                "hash": -1854333397,
              },
              "hash": -1854333397,
            },
            "hash": -1854333397,
          },
          "hash": -1854333397,
        },
        "hash": -1854333397,
      },
      "hash": -1854333397,
    },
    "hash": -1854333397,
  },
  "hash": -1854333397,
}
`;

exports[`insertIntoMerkleTree > node1TimestampStart then node2Timestamp2022 1`] = `
{
  "1": {
    "0": {
      "0": {
        "0": {
          "0": {
            "0": {
              "0": {
                "0": {
                  "0": {
                    "0": {
                      "0": {
                        "0": {
                          "0": {
                            "0": {
                              "0": {
                                "0": {
                                  "hash": -1854333397,
                                },
                                "hash": -1854333397,
                              },
                              "hash": -1854333397,
                            },
                            "hash": -1854333397,
                          },
                          "hash": -1854333397,
                        },
                        "hash": -1854333397,
                      },
                      "hash": -1854333397,
                    },
                    "hash": -1854333397,
                  },
                  "hash": -1854333397,
                },
                "hash": -1854333397,
              },
              "hash": -1854333397,
            },
            "hash": -1854333397,
          },
          "hash": -1854333397,
        },
        "hash": -1854333397,
      },
      "hash": -1854333397,
    },
    "hash": -1854333397,
  },
  "2": {
    "2": {
      "2": {
        "0": {
          "2": {
            "2": {
              "1": {
                "2": {
                  "2": {
                    "2": {
                      "0": {
                        "0": {
                          "1": {
                            "1": {
                              "2": {
                                "0": {
                                  "hash": -2002249084,
                                },
                                "hash": -2002249084,
                              },
                              "hash": -2002249084,
                            },
                            "hash": -2002249084,
                          },
                          "hash": -2002249084,
                        },
                        "hash": -2002249084,
                      },
                      "hash": -2002249084,
                    },
                    "hash": -2002249084,
                  },
                  "hash": -2002249084,
                },
                "hash": -2002249084,
              },
              "hash": -2002249084,
            },
            "hash": -2002249084,
          },
          "hash": -2002249084,
        },
        "hash": -2002249084,
      },
      "hash": -2002249084,
    },
    "hash": -2002249084,
  },
  "hash": 433128623,
}
`;

exports[`insertIntoMerkleTree > node2Timestamp2022 1`] = `
{
  "2": {
    "2": {
      "2": {
        "0": {
          "2": {
            "2": {
              "1": {
                "2": {
                  "2": {
                    "2": {
                      "0": {
                        "0": {
                          "1": {
                            "1": {
                              "2": {
                                "0": {
                                  "hash": -2002249084,
                                },
                                "hash": -2002249084,
                              },
                              "hash": -2002249084,
                            },
                            "hash": -2002249084,
                          },
                          "hash": -2002249084,
                        },
                        "hash": -2002249084,
                      },
                      "hash": -2002249084,
                    },
                    "hash": -2002249084,
                  },
                  "hash": -2002249084,
                },
                "hash": -2002249084,
              },
              "hash": -2002249084,
            },
            "hash": -2002249084,
          },
          "hash": -2002249084,
        },
        "hash": -2002249084,
      },
      "hash": -2002249084,
    },
    "hash": -2002249084,
  },
  "hash": -2002249084,
}
`;

exports[`receiveTimestamp > TimestampDuplicateNodeError 1`] = `
{
  "_id": "Either",
  "_tag": "Left",
  "left": {
    "_tag": "TimestampDuplicateNodeError",
    "node": "0000000000000001",
  },
}
`;

exports[`receiveTimestamp > should fail with clock drift 1`] = `
{
  "_id": "Either",
  "_tag": "Left",
  "left": {
    "_tag": "TimestampDriftError",
    "next": 860934720001,
    "now": 860934420000,
  },
}
`;

exports[`receiveTimestamp > should fail with clock drift 2`] = `
{
  "_id": "Either",
  "_tag": "Left",
  "left": {
    "_tag": "TimestampDriftError",
    "next": 860934720001,
    "now": 860934420000,
  },
}
`;

exports[`receiveTimestamp > wall clock is later than both the local and remote timestamps 1`] = `
{
  "counter": 0,
  "millis": 860934420001,
  "node": "0000000000000001",
}
`;

exports[`receiveTimestamp > wall clock is somehow behind > for the same timestamps millis, we take the bigger counter 1`] = `
{
  "counter": 2,
  "millis": 860934420001,
  "node": "0000000000000001",
}
`;

exports[`receiveTimestamp > wall clock is somehow behind > for the same timestamps millis, we take the bigger counter 2`] = `
{
  "counter": 2,
  "millis": 860934420001,
  "node": "0000000000000001",
}
`;

exports[`receiveTimestamp > wall clock is somehow behind > local millis is later than remote 1`] = `
{
  "counter": 1,
  "millis": 860934420002,
  "node": "0000000000000001",
}
`;

exports[`receiveTimestamp > wall clock is somehow behind > remote millis is later than local 1`] = `
{
  "counter": 1,
  "millis": 860934420002,
  "node": "0000000000000001",
}
`;

exports[`sendTimestamp > should fail with clock drift 1`] = `
{
  "_id": "Either",
  "_tag": "Left",
  "left": {
    "_tag": "TimestampDriftError",
    "next": 860934720001,
    "now": 860934420000,
  },
}
`;

exports[`sendTimestamp > should fail with counter overflow 1`] = `
{
  "_id": "Either",
  "_tag": "Left",
  "left": {
    "_tag": "TimestampCounterOverflowError",
  },
}
`;

exports[`sendTimestamp > should send monotonically with a monotonic clock 1`] = `
{
  "counter": 0,
  "millis": 860934420001,
  "node": "0000000000000000",
}
`;

exports[`sendTimestamp > should send monotonically with a regressing clock 1`] = `
{
  "counter": 1,
  "millis": 860934420001,
  "node": "0000000000000000",
}
`;

exports[`sendTimestamp > should send monotonically with a stuttering clock 1`] = `
{
  "counter": 1,
  "millis": 860934420000,
  "node": "0000000000000000",
}
`;
